open-banking-accelerator/components/com.wso2.openbanking.accelerator.identity/src/main/java/com/wso2/openbanking/accelerator/identity/app2app/App2AppAuthenticator.java

```diff
@@ -0,0 +1,214 @@
+import org.wso2.carbon.identity.application.authentication.framework.exception.AuthenticationFailedException;
+import org.wso2.carbon.identity.application.authentication.framework.model.AuthenticatedUser;
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.DeviceHandler;
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.exception.PushDeviceHandlerClientException;
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.exception.PushDeviceHandlerServerException;
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.impl.DeviceHandlerImpl;
+import org.wso2.carbon.user.api.UserRealm;
+import org.wso2.carbon.user.api.UserStoreException;
+
+import java.text.ParseException;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * App2App authenticator for authenticating users from native auth attempt.
+ */
+public class App2AppAuthenticator extends AbstractApplicationAuthenticator
+        implements FederatedApplicationAuthenticator {
+
+    private static final Log log = LogFactory.getLog(App2AppAuthenticator.class);
+    private static final long serialVersionUID = -5439464372188473141L;
+    private static DeviceHandler deviceHandler;
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getName() {
+
+        return App2AppAuthenticatorConstants.AUTHENTICATOR_NAME;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getFriendlyName() {
+
+        return App2AppAuthenticatorConstants.AUTHENTICATOR_FRIENDLY_NAME;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void processAuthenticationResponse(HttpServletRequest httpServletRequest,
+                                                 HttpServletResponse httpServletResponse,
+                                                 AuthenticationContext authenticationContext)
+            throws AuthenticationFailedException {
+
+        authenticationContext.setCurrentAuthenticator(App2AppAuthenticatorConstants.AUTHENTICATOR_NAME);
+        String jwtString =
+                httpServletRequest.getParameter(App2AppAuthenticatorConstants.DEVICE_VERIFICATION_TOKEN_IDENTIFIER);
+
+        try {
+            SignedJWT signedJWT = JWTUtils.getSignedJWT(jwtString);
+            DeviceVerificationToken deviceVerificationToken = new DeviceVerificationToken(signedJWT);
+            //Extracting deviceId and loginHint is necessary to retrieve the public key
+            String loginHint = deviceVerificationToken.getLoginHint();
+            String deviceID = deviceVerificationToken.getDeviceId();
+
+            //Checking whether deviceId and loginHint present in passed jwt
+            if (StringUtils.isBlank(loginHint) || StringUtils.isBlank(deviceID)) {
+                throw new AuthenticationFailedException(App2AppAuthenticatorConstants.REQUIRED_PRAMAS_MISSING_MESSAGE);
```

Comment:
Let's add a debug line here

Reply:
Resolved in [260e508](https://github.com/wso2/financial-open-banking/pull/67/commits/260e508da50ed60b02dc3fd2f2d27107b59691d2).